Add ability to parse incomplete XML

liberza opened the pull request:
Addresses issue https://github.com/qcam/saxy/issues/21

qcam commented:
@liberza Many thanks for implementation ❤️. The code looks great from my first glance but I'm out of town for a couple of days. I will look into it more thoroughly when I'm back.

qcam commented:
@liberza Thanks again for implementing this. I have a few suggestions on the API.

Instead of exposing the context function, we could wrap it inside an continuation state like `%Saxy.Continuation{context_fun: context_fun}`. This way the `init` and `finish` functions introduced in Saxy could be encapsulated into `Saxy.Continuation` module, then `Saxy` module only introduces  `parse_continuation` which really does the parsing work.

```ex
cont = Saxy.Continuation.new(handler, init_state, options)
{:ok, cont} = Saxy.parse_continuation(data, cont)
{:ok, cont} = Saxy.parse_continuation(data, cont)
...
{:ok, user_state} = Saxy.Continuation.finish(cont)
```

WDYT?

liberza commented:
The only concern I have with splitting it up that way is that it spreads related functionality across two modules. Maybe all three new functions should go in the Continuation module?

qcam commented:
@liberza 👍 yeah that sounds great.

liberza commented:
Alrighty, I broke out the functionality to the `Saxy.Parser.Partial` module, updated the tests, and updated the readme. I didn't wrap the context in another map yet, since I was able to encapsulate the functions in a new module without going that route. If you want I can make that change too, but since the context is supposed to be opaque to the user, it didn't seem to make much of a difference.

New usage looks like this:
```elixir
alias Saxy.Parser.Partial

xml = """
<?xml version=1.0' ?>
<foo bar=value'>
</foo>
"""
split_xml = String.split(xml, "\n")

{:ok, context} = Partial.init(MyEventHandler, initial_state)
{:ok, context} = Partial.parse(Enum.at(split_xml, 0), context)
{:ok, context} = Partial.parse(Enum.at(split_xml, 1), context)
{:ok, context} = Partial.parse(Enum.at(split_xml, 2), context)
{:ok, state} = Partial.finish(context)
```

qcam commented on README.md:
```diff
@@ -94,6 +94,25 @@ stream = File.stream!("/path/to/file") |> Stream.filter(&(&1 != "\n"))
+{:ok, context} = Saxy.parse_partial(Enum.at(split_xml, 0), context)
+{:ok, context} = Saxy.parse_partial(Enum.at(split_xml, 1), context)
+{:ok, context} = Saxy.parse_partial(Enum.at(split_xml, 2), context)
+{:ok, state} = Saxy.parse_partial_done(context)
```
Could you also update the README example?

liberza replied:
Thanks, I had updated the README example but had to merge it.

qcam commented on lib/saxy.ex:
```diff
@@ -185,7 +185,7 @@ defmodule Saxy do
     end
   end
-
+      
```
```suggestion

```

qcam commented:
Thanks again for the amazing work!
